docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,187 @@
+
+### Goals:
+The goals of this proposal are 
+* Define new stable driver API.
+* API should be versioned and backward compatible.
+* API should be extensible to accommodate future enhancements.
+* Add support for dynamic provisioning and centralized attach-detach controller and
+* The biggest advantage of flex volume is it’s simplicity. It is just an adapter around the internal API. We will keep it simple in the new design and try to follow the same adapter pattern.
+
+### Non Goals:
+* This proposal does not the address the poor installation model of existing model. It is covered in another PR #32550.
+
+## Terminology:
+**Plugin**: refers to flex volume plugin.
+
+**Plugin Driver API**: API calls defined by flex volume plugin.
+
+**Plugin Driver**: Out-of-tree driver/plugin which implements plugin driver API.
+
+**In-tree plugin**: Plugins which are included in Kubernetes project/source tree.
+
+## API:
+The existing flex volume plugin driver API is in the alpha phase and are subject to change between Kubernetes versions. This section proposes a stable API for plugin driver writers to work with.
+
+### Proposal:
+Proposes fewer APIs with stable request & response API objects.
+
+###### Pros:
+* Has standard request & response API objects.
+* Supports versioning.
+* Uses gRPC.
+* Fewer calls.
+* Simple and clean interface.
+
+###### Cons:
+* New API objects.
+
+#### APIs
+
+```protobuf
+
+
+// Capabilities of the driver
+message DriverCapabilities {
+    enum ReadWriteMode {
+        READ_ONLY = 0;
+        READ_WRITE_ONCE = 1;
+        READ_WRITE_MANY = 2;
+    }
+	ReadWriteMode supported_read_write_mode = 1;
+    enum AttachmentPolicy {
+        NONE = 0;  // Driver does not support attach/detach.
+        LOCAL = 1; // Driver supports local volume attachments from kubelet.
+        REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+    }
+    AttachmentPolicy supported_attachment_policy = 2;
+    enum Capabilities {
+        DYNAMIC_PROVISIONING = 0;
+        CUSTOM_MOUNT = 1;
+        SELINUX = 2,
+        OWNERSHIP_MANAGEMENT = 3;
+        METRICS = 4;
+    }
+    repeated Capabilities supported_capabilities = 3;
+}
+
+message ProbeDriverRequest {}
+
+message ProbeDriverResponse {
+    DriverCapabilities capabilities = 1;
+    repeated string options = 2;
+}
+
+service FlexVolumeService {
+    rpc ProbeDriver(ProbeDriverRequest) returns (ProbeDriverResponse) {}
+    rpc Create(CreateRequest) returns (CreateResponse) {}
+    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+    rpc Attach(AttachRequest) returns (AttachResponse) {}
+    rpc Detach(DetachRequest) returns (DetachResponse) {}
+    rpc Mount(MountRequest) returns (MountResponse) {}
+    rpc Unmount(UnmountRequest) returns (UnmountResponse) {}
+    rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}
+}
+
+// Volume source.
+message Spec {
```

Comment:
I am not sure, can you simply send whole `FlexVolumeSource` instead of its items here?